d2/src/main/java/com/linkedin/d2/xds/XdsClient.java

```diff
@@ -0,0 +1,280 @@
+package com.linkedin.d2.xds;
+
+import com.google.protobuf.Value;
+import io.envoyproxy.envoy.config.endpoint.v3.LocalityLbEndpoints;
+import io.envoyproxy.envoy.config.route.v3.VirtualHost;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+
+public abstract class XdsClient
+{
+  /**
+   * Watcher interface for a single requested xDS resource.
+   */
+  interface ResourceWatcher
+  {
+    // TODO: Handle onError and onResourceNotExist
+  }
+
+  interface LdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(LdsUpdate update);
+  }
+
+  interface RdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(RdsUpdate update);
+  }
+
+  interface CdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(CdsUpdate update);
+  }
+
+  interface EdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(EdsUpdate update);
+  }
+
+  interface ResourceUpdate
```

Comment:
Maybe add a type() / opType() method that returns the type of the update operation: add / removal / patch / etc. Then we can extend the update handler in the future iterations.

Reply:
This will be handled by the ResourceWatcher interface: https://github.com/grpc/grpc-java/blob/master/xds/src/main/java/io/grpc/xds/XdsClient.java#L399